Fix non-deterministic coverage of test DoS_mapOrphans

davereikher opened the pull request:
This pull request proposes a solution to make the test `DoS_mapOrphans` in denialofservice_tests.cpp have deterministic coverage.

The `RandomOrphan` function in denialofservice_tests.cpp and the implicitly called function `ecdsa_signature_parse_der_lax` in pubkey.cpp were causing the non-deterministic test coverage.

In the former, if a random orphan was selected the index of which is bigger than the max. orphan index in `mapOrphanTransactions`, the last orphan was returned from `RandomOrphan`. If the random number generated was never large enough, this condition would not be fulfilled and the corresponding branch wouldn't run. The proposed solution is to force one of the 50 dependant orphans to depend on the last orphan in `mapOrphanTransactions` using the newly introduced function `OrphanByIndex` (and passing it a large uint256), forcing this branch to run at least once.

In the latter, if values for ECDSA `R` or `S` (or both) had no leading zeros, some code would not be executed. The solution was to find a constant signature that would be comprised of `R` and `S` values with leading zeros and calling `CPubKey::Verify` at the end of the test with this signature forcing this code to always run at least once at the end even if it hadn't throughout the test.

To test that the coverage is (at least highly likely) deterministic, I ran 

`contrib/devtools/test_deterministic_coverage.sh denialofservice_tests/DoS_mapOrphans 1000`

and the result was deterministic coverage across 1000 runs.

Also - removed denialofservice_tests test entry from the list of non-deterministic tests in the coverage script.

davereikher committed:
Make test DoS_mapOrphans deterministic

The RandomOrphan function and the function ecdsa_signature_parse_der_lax
in pubkey.cpp were causing non-deterministic test coverage.

Force seed in the beginning of the test to make it deterministic.
The seed is selected carefully so that all branches of the function
ecdsa_signature_parse_der_lax are executed. Prior to this fix, the test
was exhibiting non-deterministic coverage since none of the ECDSA
signatures that were generated during the test had leading zeroes in
either R, S, or both, resulting in some branches of said function not
being executed. The seed ensures that both conditions are hit.

Removed denialofservice_tests test entry from the list of non-deterministic
tests in the coverage script.

practicalswift commented:
Concept ACK

Nice work! If you have the time: consider tackling the remaining `test_deterministic_coverage.sh` suppressions in follow-up PRs. Deterministic line coverage would be great to have! :)

practicalswift commented on src/test/denialofservice_tests.cpp:
```diff
@@ -391,7 +418,14 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
+        if (0 == i) {
+            // This block makes sure that the condition "if (it == mapOrphanTransactions.end())" in OrphanByIndex() gets called at least once.
+            // Otherwise test coverage is non-deterministic.
+            txPrev = OrphanByIndex(ArithToUint256(arith_uint256(LARGE_NUMBER)));
```
Could use `std::numeric_limits<uint64_t>::max()` instead of `LARGE_NUMBER`? :)

practicalswift commented on src/test/denialofservice_tests.cpp:
```diff
@@ -391,7 +418,14 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
     {
-        CTransactionRef txPrev = RandomOrphan();
+        CTransactionRef txPrev;
+        if (0 == i) {
```
Nit: Could avoid non-repo-idiomatic yoda notation :)

davereikher commented:
> Concept ACK
> 
> Nice work! If you have the time: consider tackling the remaining `test_deterministic_coverage.sh` suppressions in follow-up PRs. Deterministic line coverage would be great to have! :)

Thanks for the review! I addressed your remarks in ffd3953. I indeed plan to tackle those in my free time.

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -356,16 +358,39 @@ BOOST_AUTO_TEST_CASE(DoS_bantime)
+    hash.SetHex(std::string("6af516422fef8a745aff6acdcc84076c77fb2ecd72bd5711df301230ac58fdd5"));
+
+    coverage_pubkey.Verify(hash, std::vector<unsigned char>(vch_sig_str.begin(), vch_sig_str.end()));
+}
```
Looks like the wrong location for this. I fail to understand what CPubKey has to do with the orphan map or even DoS?!

davereikher replied:
Thanks for the review! While trying to understand the non-determinism in the coverage, I used the script `contrib/devtools/test_deterministic_coverage.sh`, which uses `gcovr` to show the diff in which lines were and which ones were not executed between two runs with different coverage. 

Over multiple runs of the script, the only two places where the coverage was non-deterministic were: 

- the function `RandomOrphan` in denialofservice_tests.cpp 
- lines 135-136, 147-148 in pubkey.cpp, in the function `ecdsa_signature_parse_der_lax`. 

The latter function is called implicitly from the test `DoS_mapOrphans` via `SignSignature` (`SignSignature` => `ProduceSignature` => `VerifyScript` => `EvalScript` => `CheckSig` => `VerifySignature` => `Verify` => `ecdsa_signature_parse_der_lax`). The private key which generates the signatures in the test `DoS_mapOrphans` is randomly generated, so the `R` and `S` values of the ECDSA signatures for the inputs of the dependant orphans most of the time do not have leading zeros. The lines 135-136, 147-148 run only when there are leading zeroes in `R`, `S`, respectively. This results in those lines sometimes running and sometimes not. 

My proposed solution is to create a constant signature with leading zeroes in both `R` and `S` and force the function `ecdsa_signature_parse_der_lax` to run on that signature by calling `CPubkey::Verify` on that signature and a corresponding pubkey and hash. This forces the lines 135-136, 147-148 in pubkey.cpp to always run at least once, resulting in deterministic test coverage.

MarcoFalke replied:
I see. A smaller alternative would be to set the key to a constant?

davereikher replied:
That could be nice actually if it's acceptable. I didn't want to do it initially because I didn't want to modify the test too much and restricted myself to 'parasitic' changes.
On the other hand - if I'm not mistaken, this will not work directly, since, if we have transaction B which is using as an input an output from transaction A, part of the data used to calculate the hash of that input for B (this hash is then signed to produce `R` and `S`) is the hash of the transaction A, which contains a random element (the initial non-dependant 50 orphans are created with `tx.vin[0].prevout.hash = InsecureRand256()`), so `R` and `S` would again be random.

I guess I could use an engineered, non-random value for `tx.vin[0].prevout.hash` for one of the orphans along with an engineered fixed key. But then again, this test would loose the fuzzing feature of making a different key every run.

What do you think?

MarcoFalke replied:
`InsecureRand256` is an alias for `g_insecure_rand_ctx.rand256()`, so it can be made deterministic with a call to `SeedInsecureRand(/* deterministic */ true)`.

> this test would loose the fuzzing feature of making a different key every run.

This "fuzzing feature" is useless, as a failure could not be reproduced (when it occurs) and thus not debugged, nor fixed.

When fuzzing is desired, the seed would have to be written to a file or stdout.

davereikher replied:
I see. Then I can just set a seed which guarantees the widest coverage in the beginning of the test and that should take care of the non-determinism in both `RandomOrphan` and `ecdsa_signature_parse_der_lax`, making the code much simpler.

Still, I would argue that it's best to leave the fuzzing and write the seed to a file. That would result in an a better coverage over a long period of time. There might be some branches that run very rarely and setting the seed to a constant would prevent them from ever running in the future in the context of this test, reducing it's long-term coverage. As an example the conditions on lines 151, 157, 160 in pubkey.cpp, in `ecdsa_signature_parse_der_lax` may evaluate to false or true. Although their evaluation was deterministic throughout the 1000+ runs I tried and I haven't gone in-depth to try and understand whether they can at all run under this test's conditions or whether they run extremely rarely, but the latter option can be true for other branches I haven't checked.
Unless handling artifacts created by tests involves a lot of overhead, the only disadvantage I see to leaving the fuzzing is more lines of code in the test, some of it being irrelevant to the functionality of the test itself, but I think that is taken care of by a detailed comment explaining the logic for the extra code.

MarcoFalke replied:
Yeah, up to you if you feel strongly. My opinion is that we shouldn't mix unit tests with fuzz tests, which are designed and compiled separately, see https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md

davereikher replied:
I agree that fuzz tests should be separate and I wasn't familiar with `src/test/fuzz`. Thanks for pointing me there. However, reading through https://github.com/bitcoin/bitcoin/pull/16320, there are some comments about the positive value of non-determinism in tests (I was trying to understand if the direction is to completely eliminate non-determinism). I think that an effort should be made when writing new tests towards determinism and separating the fuzzy parts into the fuzz framework, but since this test already has non-determinism, I think I will stick with writing the seed to stdout. Separating the fuzzy from the non-fuzzy parts should perhaps be a part of another pull request.
That brings the question - there are a lot of tests that use random values. I completely agree that randomness in tests is practically useless unless the seed is known. What do you think about setting and writing the seed to stdout as part of the `BasicTestingSetup` fixture so that seeds for all tests are known? This would also prevent code duplication when multiple tests require this.

MarcoFalke replied:
We already use `BOOST_TEST_RANDOM` https://www.boost.org/doc/libs/1_71_0/libs/test/doc/html/boost_test/utf_reference/rt_param_reference/random.html, which is logged. So might as well use that as the seed for our test rng?

davereikher replied:
Sounds good. Do you suggest seeding the rng with `BOOST_TEST_RANDOM` in `BasicTestingSetup` or only in this test for now? I'm thinking the latter, since otherwise it changes the behaviour of all tests that use random numbers when running them locally and it would require adjusting the test execution scripts, as now the seed must be provided externally. For example, `test_deterministic_coverage.sh` will always show deterministic coverage when run locally unless we provide a different value of `BOOST_TEST_RANDOM` every time it's run. I feel handling those issues would be out of the scope of this pull request and may be done separately?

MarcoFalke replied:
> test_deterministic_coverage.sh will always show deterministic coverage when run locally unless we provide a different value of BOOST_TEST_RANDOM every time it's run. I feel handling those issues would be out of the scope of this pull request and may be done separately?

I think this is the way the script should work (report non-determinism even when the rng seed is pinned to a constant). As suggested by you this should be done in a separate pull request. Are you interested in working on this as well? Otherwise I will create an issue for others to pick up.

davereikher replied:
Yes, that sounds interesting, since IMO recording the seed for any test that uses random values is important. I'll check whether pinning the seed still results in non determinism between runs and based on that I'll see what can be done about it and about setting the seed to `BOOST_TEST_RANDOM` in `BasicTestingSetup`.
For now, I guess setting the seed to `BOOST_TEST_RANDOM` just inside this test would be enough?

davereikher commented:
Could somebody explain why it's not allowed to include `boost/test/unit_test_parameters.hpp`?
```
A new Boost dependency in the form of "boost/test/unit_test_parameters.hpp" appears to have been introduced:

src/test/denialofservice_tests.cpp:#include <boost/test/unit_test_parameters.hpp>

^---- failure generated from test/lint/lint-includes.sh
```
I need it to access the `BOOST_TEST_RANDOM` parameter from within a test. Is there a better way to do it?

MarcoFalke commented:
I guess, you'd have to add it like this:

```diff
diff --git a/test/lint/lint-includes.sh b/test/lint/lint-includes.sh
index d27e45a23f..c9e1d57ee1 100755
--- a/test/lint/lint-includes.sh
+++ b/test/lint/lint-includes.sh
@@ -68,6 +68,7 @@ EXPECTED_BOOST_INCLUDES=(
     boost/signals2/last_value.hpp
     boost/signals2/signal.hpp
     boost/test/unit_test.hpp
+    boost/test/unit_test_parameters.hpp
     boost/thread.hpp
     boost/thread/condition_variable.hpp
     boost/thread/mutex.hpp

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -393,6 +394,8 @@ static void ForceCoverageInPubKeyVerify()
 
 BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
 {
+    SeedInsecureRand(ArithToUint256(arith_uint256(boost::unit_test::runtime_config::get<unsigned>( boost::unit_test::runtime_config::btrt_random_seed ))));
```
I guess you could move this logic into a helper `SeedInsecureRandWithBoostRandomSeed`?

MarcoFalke commented on src/test/setup_common.h:
```diff
@@ -43,6 +45,16 @@ static inline void SeedInsecureRand(bool deterministic = false)
+
+static inline void SeedInsecureRandWithBoostRandomSeed()
+{
+    SeedInsecureRand(ArithToUint256(arith_uint256(boost::unit_test::runtime_config::get<unsigned>( boost::unit_test::runtime_config::btrt_random_seed ))));
```
not sure how stable that boost interface is. They seem to change it every couple of versions

davereikher replied:
Is there a better way to access this value? Maybe directly access the environment variable using std::getenv?

davereikher commented:
@MarcoFalke I see that the changes I made regarding setting the seed to `BOOST_TEST_RANDOM` are causing some trouble - some Travis builds fail while succeeding for me locally so I'm not sure how to debug that as well as the fact that the boost interface might change and I also feel that this is going a bit out of scope for this pull request which aims to make DoS_mapOrphans have deterministic coverage.
Would it be ok if I leave the test with a random, unknown seed for now (that was the case before anyway) and just leave the changes that make the coverage deterministic for this test? I will then attempt to tackle the problem of setting the seed to `BOOST_TEST_RANDOM` in `BasicTestingSetup` in a separate pull request?
Overall, I think this would still add positive value, as this test would have deterministic coverage, as opposed to before, with all else being equal.

practicalswift commented:
ACK ffd3953757051ac50cea4129edbd9e9d5be39d0f -- diff looks correct

davereikher commented:
How can I remove the "Waiting for author" label?

MarcoFalke commented:
Have you seen #16978, which should fix the reproducibility issue

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -356,16 +358,39 @@ BOOST_AUTO_TEST_CASE(DoS_bantime)
+    hash.SetHex(std::string("6af516422fef8a745aff6acdcc84076c77fb2ecd72bd5711df301230ac58fdd5"));
+
+    coverage_pubkey.Verify(hash, std::vector<unsigned char>(vch_sig_str.begin(), vch_sig_str.end()));
+}
```
Still don't like this, maybe it can be removed after https://github.com/bitcoin/bitcoin/pull/16878#issuecomment-538098710 ?

davereikher replied:
Actually, I ran the `DoS_mapOrphans` test to test the initial version of #16978 and even though the seed was the same between different runs, I still got non-deterministic behaviour for this test. The non-determinism originated from the size of the `mapOrphanTransactions` map. It was different for each run. I didn't dive into the source code to understand why exactly that is but I suspect that it's due to concurrency issues, since the seed was fixed between runs (I printed some `uint256`'s generated by `InsecureRand256()` to make sure and they were the same between runs),  unless there is another, differently seeded RNG used in the process of populating `mapOrphanTransactions`. In any case, if I understand correctly, #16978  is a great solution for setting the seed externally and for logging it, but it won't remove the non-determinism between test runs of this particular test, since I suspect that the non-determinism in this test arises due to concurrency and not due to seeding of an RNG. Though I agree that the fix is ugly, this is unfortunately the only way I can see of fixing the non-deterministic coverage issue for this particular test.

MarcoFalke replied:
Hmm, ok. Will see if other have an opinion on this.

davereikher commented:
@laanwj  This PR is waiting for additional reviews, so I don't think this should be labeled as 'Waiting for author'?

jonatack commented on src/test/denialofservice_tests.cpp:
```diff
@@ -16,6 +16,8 @@
 #include <util/time.h>
 #include <validation.h>
+#include <arith_uint256.h>
+#include <pubkey.h>
```
nit: sort

jonatack commented on src/test/denialofservice_tests.cpp:
```diff
@@ -356,16 +358,39 @@ BOOST_AUTO_TEST_CASE(DoS_bantime)
+/** This function runs CPubkey::Verify with parameters that result in all branches of the function ecdsa_signature_parse_der_lax to run. Namely, the signature literal
+ * in vch_sig_str in the following function is comprised of R and S values which both contain leading zeroes, forcing some branches of said function to run which
+ * would otherwise not. This function is called at the end of the DoS_mapOrphans test to force deterministic coverage.
+ */
```
suggestion
```diff
-/** This function runs CPubkey::Verify with parameters that result in all branches of the function ecdsa_signature_parse_der_lax to run. Namely, the signature literal
- * in vch_sig_str in the following function is comprised of R and S values which both contain leading zeroes, forcing some branches of said function to run which
- * would otherwise not. This function is called at the end of the DoS_mapOrphans test to force deterministic coverage.
+/** This function runs CPubkey::Verify with parameters that result in all
+ * branches of the function ecdsa_signature_parse_der_lax to run. Namely, the
+ * signature literal in vch_sig_str in the following function is comprised of R
+ * and S values which both contain leading zeroes, forcing some branches of said
+ * function to run which would otherwise not. This function is called at the end
+ * of the DoS_mapOrphans test to force deterministic coverage.
  */
```

jonatack reviewed (commented):
ACK ffd3953757051ac50cea4129edbd9e9d5b tested rebased on current master

Feel free to ignore the nits below.

davereikher commented:
Thanks very much for the review @jonatack ! I addressed the nits in  d90308b.
Could somebody please re-run the travis job? Looks like it timed out on ARM.

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -391,7 +419,14 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
+        if (i == 0) {
+            // This block makes sure that the condition "if (it == mapOrphanTransactions.end())" in OrphanByIndex() gets called at least once.
+            // Otherwise test coverage is non-deterministic.
+            txPrev = OrphanByIndex(ArithToUint256(arith_uint256(std::numeric_limits<uint64_t>::max())));
```
instead of hardcoding special cases, I'd prefer if the txids were just derived from a deterministic fast random context. If you want, you can explicitly seed it so that the condition is hit at least once.

davereikher replied:
Ok, I can do that. The RNG can indeed be seeded so that the condition is hit at least once. However, the `ForceCoverageInPubKeyVerify` function must remain, since the non-determinism is not caused by different seeds, but is there due to concurrency issues (please refer to my comment [here](https://github.com/bitcoin/bitcoin/pull/16878?_pjax=%23js-repo-pjax-container#discussion_r331741093) for some details). I therefore don't see another way to remove the non-determinism other than this. Is that ok?

MarcoFalke replied:
Generally the tests run in a single thread, so there shouldn't be any concurrency issues. Are you sure the non-determinism is not caused by the random private key, which can also be hardcoded?

davereikher replied:
Sorry for the long delay, you are right, the additional non-determinism was caused by the fact that the private key is generated by another RNG and not by concurrency issues. I removed the function I added to force coverage and forced the key to be generated by the same RNG for this test. I seeded it so that all branches of the function `ecdsa_signature_parse_der_lax` run. However, after a few dozen times of running this test with the script `contrib/devtools/test_deterministic_coverage.sh`, there was another place where the coverage was not deterministic, in the function `Loop()` in `checkqueue.h`. I'm looking into it.

maflcko reviewed (changes requested):
the changes are incomplete because they don't print the seed, so the "fuzzing feature" doesn't add any value because if a test failed, it would be impossible to reproduce without the seed.

I think that fuzzing should be done in the fuzz tests and not unit tests and my suggestion would be to make the test deterministic here.

davereikher commented:
I'm running this test multiple times with the `contrib/devtools/test_deterministic_coverage.sh` script and out of 590 runs it detected non-deterministic coverage twice, in a seemingly unrelated location. This line https://github.com/bitcoin/bitcoin/blob/9a714c51dc79993506718617e383294cf6bff0b4/src/checkqueue.h#L99 very rarely doesn't run. I'm thinking that maybe some tests which are not in the list of non-deterministic tests in this script: https://github.com/bitcoin/bitcoin/blob/9a714c51dc79993506718617e383294cf6bff0b4/contrib/devtools/test_deterministic_coverage.sh#L16 might also have non-deterministic behavior if run enough times with the script.
@practicalswift Can you say how many times the script ran for each test before it was considered to be deterministic? I guess there is some threshold number of runs such that if non-deterministic behavior doesn't appear during those runs then this test is considered to be deterministic 'enough'. Is 2 out of 590 runs above or below that threshold?

MarcoFalke commented:
You don't need to solve all issues. As long as something is an improvement, it should be good to go

practicalswift commented:
@davereikher 

> @practicalswift Can you say how many times the script ran for each test before it was considered to be deterministic? I guess there is some threshold number of runs such that if non-deterministic behavior doesn't appear during those runs then this test is considered to be deterministic 'enough'. Is 2 out of 590 runs above or below that threshold?

Personally I think 2 out of 590 runs is a great improvement compared to the current state of things: great job! :)

Thanks for tackling testing non-determinism!

davereikher commented:
> @davereikher
> 
> > @practicalswift Can you say how many times the script ran for each test before it was considered to be deterministic? I guess there is some threshold number of runs such that if non-deterministic behavior doesn't appear during those runs then this test is considered to be deterministic 'enough'. Is 2 out of 590 runs above or below that threshold?
> 
> Personally I think 2 out of 590 runs is a great improvement compared to the current state of things: great job! :)
> 
> Thanks for tackling testing non-determinism!

Thanks, so do you think this test should be removed from the list of non-deterministic tests then?

davereikher commented:
Ok, I think it's done. As [stated](https://github.com/bitcoin/bitcoin/pull/16878#issuecomment-659194950), there is still some non-determinism, but it occurs once every several hundred rounds in an unrelated location. I'm assuming it's below the threshold of non-determinism, so I removed this test from the list of non-deterministic tests in `contrib/devtools/test_deterministic_coverage.sh`.

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -314,10 +316,27 @@ static CTransactionRef RandomOrphan()
+    do {
+        keydata = g_insecure_rand_ctx.randbytes(32);
+        key.Set(keydata.data(), keydata.data() + keydata.size(), /*fCompressedIn*/ true);
+    } while (!key.IsValid());
```
nit: This can simply say
```suggestion
    assert(key.IsValid());
```

davereikher replied:
Wouldn't it fail the test if the generated key is invalid instead of retrying?

MarcoFalke replied:
Yes, but that should never happen, because the test is deterministic

maflcko reviewed (approved):
ACK

MarcoFalke commented:
ACK 4455949d6f0218b40d33d7fe6de6555f8f62192f

jonatack reviewed (commented):
ACK 4455949d6f0218b40d33d7fe6de6555f8f62192f

Oops, too late for the merge.
